apps/emqx_retainer/src/emqx_retainer.erl

```diff
@@ -311,3 +324,13 @@ condition(Ws) ->
         false -> Ws1;
         _ -> (Ws1 -- ['#']) ++ '_'
     end.
+
+clean_by_clientid(ClientId) ->
+    MsHd = #retained{topic = '$1', msg = '$2', expiry_time = '_'},
+    Ms = [{MsHd, [{'=:=', {element, 4, '$2'}, ClientId}], ['$1']}],
+    {atomic, _} = mnesia:transaction(
+        fun() ->
+            Keys = mnesia:select(?TAB, Ms, write),
+            lists:foreach(fun(Key) -> mnesia:delete({?TAB, Key}) end, Keys)
+        end),
+    ok.
```

Comment:
this may cause table lock.
given the scan is quite expensive.
please change it to dirty delete.